engine/src/main/java/org/apache/apex/engine/plugin/DefaultApexPluginDispatcher.java

```diff
@@ -87,18 +87,24 @@ public void rejectedExecution(Runnable r, ThreadPoolExecutor executor)
     };
 
     executorService = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS,
-        blockingQueue, new NameableThreadFactory("PluginExecutorThread"), rejectionHandler);
+        blockingQueue, new NameableThreadFactory("PluginExecutorThread", true), rejectionHandler);
   }
 
   @Override
   protected void serviceStop() throws Exception
   {
-    executorService.shutdownNow();
-    executorService.awaitTermination(10, TimeUnit.SECONDS);
-    if (!executorService.isTerminated()) {
-      LOG.warn("Executor service still active for plugins");
+    if (executorService != null) {
+      executorService.shutdown();
+      boolean terminated = false;
+      try {
+        terminated = executorService.awaitTermination(10, TimeUnit.SECONDS);
+      } finally {
+        if (!terminated && !executorService.isTerminated()) {
```

Comment:
Depends on the executorService implementation and possibly requires some synchronization. As awaitTermination() already returns the state, I can remove the extra call to isTerminated() if you feel strongly about it.